### PR TITLE
First version of the docker deployment image

### PR DESCRIPTION
**Pull request by fischerbenjamin**
I created a first version of the docker image we want to use as the deployment.
I moved all the configs into a separate folder and updated the hard coded paths inside the crawler according to it.
Also changed the default ports to be consistent.
If you have any questions, just comment below :)

### CODE REVIEW

File: docker/deployment/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    scripts/database.sh
+USER root
+# Install crawler dependencies
+COPY crawler/requirements crawler/requirements
+RUN pip3 install -r crawler/requirements/base.txt
+COPY crawler crawler
+# Copy server artifacts
+COPY --from=server-build metadatahub/server/target/metadata-hub-server-application-fat.jar server/
+COPY --from=server-build metadatahub/web-ui web-ui
+# Copy configs, export ports and set entrypoint and default arguments via cmd
```
Comment (knotenpunkt) [inline docker/deployment/Dockerfile:46]: we only need the ....fat.jar.
maven includes into this with the help from webpack all necassary files.

maven:
```
 <configuration>
                    <workingDirectory>../web-ui/</workingDirectory>
                    <nodeVersion>v12.16.3</nodeVersion>
                    <npmVersion>6.14.5</npmVersion>
 </configuration>
```
[...]
``` 
<execution>
                        <id>webpack-build</id>
                        <goals><goal>npm</goal></goals>
                        <configuration>
                            <arguments>run dev-build</arguments>
                        </configuration>
</execution>
```

in the package.json-file:
```
    "scripts": {
        "build": "webpack",
        "dev-build": "webpack -o ../server/src/main/resources/static-files/bundle.js",
        "runtime-build": "webpack -o ../server/target/classes/static-files/bundle.js",
       "test": "echo \"Error: no test specified\" && exit 1"
    },
```

so maven calls the dev-build (`<arguments>run dev-build</arguments>`). This bundle the files (there are more bundel.js which are then outputet , main.bundle.js, bundle.js, something like that). This goes to the ressource-directory. And in the final, if maven could compile the java-code, all of that goes to the jar-file. 
So its only necassary, that maven can access `  <workingDirectory>../web-ui/</workingDirectory>`

Maybe its a good idea to cache the **node/ and node_modules/**, so the build could be faster

Comment (fischerbenjamin) [reply docker/deployment/Dockerfile:46]: Thanks a lot, I'll work on that :)

**Comment by fischerbenjamin** [comment]
So, I updated the docker versions. It should work now with the new config of the server.
Furthermore, the branches _master_ and _develop_ are automatically pushed to DockerHub. I also included a description there.

**Comment by fischerbenjamin** [comment]
> configuration of the database, e.g use other db, user, password, etc.?

This is something that cannot be configured that easily, because the initialization of the database happens during build time. This is done by the hard-coded .sql files. Thus, modified .sql scripts have to be provided that contain the corresponding user, database and password. Admin user/password could be configured via the environment.json file.

> how can we reset the database data?

Either, just delete the volume and re-create it with the same name, or use another empty one

> if we use there always the old database-data what happens if the schemata changes, or we want to use a other initial-data-set?

If you change the schema by modifying one of the .sql files in the *database* directory, simply rebuild the image, the changes should get applied. If you provide additional .sql files or delete any of them, the *Dockerfile* has to be updated as well.

> what is for now the rule, with that the container works

You can take a look at the README in *docker/deployment* or here [here](https://hub.docker.com/r/amosproject2/metadatahub)

**Comment by knotenpunkt** [comment]
> > configuration of the database, e.g use other db, user, password, etc.?
> 
> This is something that cannot be configured that easily, because the initialization of the database happens during build time. This is done by the hard-coded .sql files. Thus, modified .sql scripts have to be provided that contain the corresponding user, database and password. Admin user/password could be configured via the environment.json file.
> 
maybe you misunderstood me here: i meant, that the user should have the option to link to a ready-one-full-initalized database somewhere extern. Its not the task from docker or at build-time to create these database. I you want to use a extern database, then you have to set up on your own.

the environment.json is i think included in the docker-container, so i cant easy rewrite it? So my idea was, that we also should/could inject (in the future) some config-stuff from the outside.


> > how can we reset the database data?
> 
> Either, just delete the volume and re-create it with the same name, or use another empty one
> 
> > if we use there always the old database-data what happens if the schemata changes, or we want to use a other initial-data-set?
> 
> If you change the schema by modifying one of the .sql files in the _database_ directory, simply rebuild the image, the changes should get applied. If you provide additional .sql files or delete any of them, the _Dockerfile_ has to be updated as well.
> 

so that means, every-time i rebuild the image, the old database-data would be erased?
regardless of whether, if i changed the schemata or not?

> > what is for now the rule, with that the container works
> 
> You can take a look at the README in _docker/deployment_ or here [here](https://hub.docker.com/r/amosproject2/metadatahub)

**Comment by fischerbenjamin** [comment]
> so that means, every-time i rebuild the image, the old database-data would be erased?
regardless of whether, if i changed the schemata or not?

Not if you use a volume to store the actualy database. That volume lives on the host, so it's indepent of the actual image. If you don't use a volume, the data is stored in the container.

> the environment.json is i think included in the docker-container, so i cant easy rewrite it? So my idea was, that we also should/could inject (in the future) some config-stuff from the outside.

Hm, I don't know if it is necessary to change for example the database port during runtime. If you have a running container that you would like to reconfigure, then you would have to stop the services anyway, change the config file and restart them. I don't think that makes sense. I'm not quite sure how docker run works exactly, but in this case it would be a better idea to somehow pass the environment.config to docker run and mount the directory into the container where it is located on the host. 

> maybe you misunderstood me here: i meant, that the user should have the option to link to a ready-one-full-initalized database somewhere extern. Its not the task from docker or at build-time to create these database. I you want to use a extern database, then you have to set up on your own.

I think this should work with the volumes

**Comment by knotenpunkt** [comment]
> > so that means, every-time i rebuild the image, the old database-data would be erased?
> > regardless of whether, if i changed the schemata or not?
> 
> Not if you use a volume to store the actualy database. That volume lives on the host, so it's indepent of the actual image. If you don't use a volume, the data is stored in the container.
> 

yeah that i mean^^
But then if we had a change in our schemata, the new schemata is not pushed into the existing db.
("Wie auch", yeah in that case we would need a db-updater, and thats - i think - a overkill)
But maybe we need there something, what checks the schemata versions, and if it changes, print the user that the old schemata that lives in the volume doesnt work anymore.


> > the environment.json is i think included in the docker-container, so i cant easy rewrite it? So my idea was, that we also should/could inject (in the future) some config-stuff from the outside.
> 
> Hm, I don't know if it is necessary to change for example the database port during runtime. If you have a running container that you would like to reconfigure, then you would have to stop the services anyway, change the config file and restart them. I don't think that makes sense. I'm not quite sure how docker run works exactly, but in this case it would be a better idea to somehow pass the environment.config to docker run and mount the directory into the container where it is located on the host.
> 

ah ok^^, no it was not meant to change it on runtime^^
But i thought having something like that the user dont have to dive into the docker-code-base; so that we have there only our docker-container, and a extern config, and a extern-volume

(But for our develop-stuff its ok i think; it was more for productive-system thought)


> > maybe you misunderstood me here: i meant, that the user should have the option to link to a ready-one-full-initalized database somewhere extern. Its not the task from docker or at build-time to create these database. I you want to use a extern database, then you have to set up on your own.
> 
> I think this should work with the volumes

yeah, but if the db lives on a other machine, i mean a whole postgres-installation, that could be on a other server on the other side of the world

**Comment by fischerbenjamin** [comment]
> But then if we had a change in our schemata, the new schemata is not pushed into the existing db.
("Wie auch", yeah in that case we would need a db-updater, and thats - i think - a overkill)
But maybe we need there something, what checks the schemata versions, and if it changes, print the user that the old schemata that lives in the volume doesnt work anymore.

That is true. I have no idea tbh for an approach that is simple enough & fast to implement for us to consider really doing it yet, but if you have some idea, please let me know :) I'll think about it!

> But i thought having something like that the user dont have to dive into the docker-code-base; so that we have there only our docker-container, and a extern config, and a extern-volume
(But for our develop-stuff its ok i think; it was more for productive-system thought)

Ah okay, I see. We can have a chat on that if you want to, because it would be a lot to write right now^^

> yeah, but if the db lives on a other machine, i mean a whole postgres-installation, that could be on a other server on the other side of the world

I think this should work if you use the host network for the container and specify it in the config and rebuild the images with it. Or you use the compose variant, because both server and crawler only connect via host/port to the database and don't rely on some filesystem connectivity.

**Comment by knotenpunkt** [comment]
> > But then if we had a change in our schemata, the new schemata is not pushed into the existing db.
> > ("Wie auch", yeah in that case we would need a db-updater, and thats - i think - a overkill)
> > But maybe we need there something, what checks the schemata versions, and if it changes, print the user that the old schemata that lives in the volume doesnt work anymore.
> 
> That is true. I have no idea tbh for an approach that is simple enough & fast to implement for us to consider really doing it yet, but if you have some idea, please let me know :) I'll think about it!
> 

maybe we could solve that on application-level:
so we add into the db one more table called version with only one row called data.
we also add into the schemata.sql-file not into the data.sql-file a insert-command to this version-table, that inserts a hard-coded integer-value into the data-row, which then is our version.

**-> so now the version-data here is not anymore a db-data-entry but more a component of the schemata itself**

the Crawler and the Java-Application check this value on its start-time. If it is wrong, both would print a error-message, that the databse, they want to work with, is not that one, what they excpect. 

Maybe this could be a solution.






> > But i thought having something like that the user dont have to dive into the docker-code-base; so that we have there only our docker-container, and a extern config, and a extern-volume
> > (But for our develop-stuff its ok i think; it was more for productive-system thought)
> 
> Ah okay, I see. We can have a chat on that if you want to, because it would be a lot to write right now^^
> 
> > yeah, but if the db lives on a other machine, i mean a whole postgres-installation, that could be on a other server on the other side of the world
> 
> I think this should work if you use the host network for the container and specify it in the config and rebuild the images with it. Or you use the compose variant, because both server and crawler only connect via host/port to the database and don't rely on some filesystem connectivity.

yeah i think for this is a chat better, but we dont have to do this for now, cause the stuff here is not really important for the PR.

**Comment by KhangNguyenNgoc** [comment]
I think it looks good and works

**Pull request merged**